### PR TITLE
Override completion with in parameter should omit IsReadOnly attribute

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/24490
Fixes https://github.com/dotnet/roslyn/issues/24615

Update: the PR now filters the `IsReadOnly` and `IsByRefLike` attributes in the compiler `GetAttributes()` API.

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/OverrideCompletionProviderTests.cs [hunk 1]
```diff
@@ -777,6 +777,23 @@ public class SomeClass : Base
             await VerifyItemExistsAsync(markup, "goo(int x, out string y)", "void Base.goo(int x, out string y)");
         }
 
+        [WpfFact, Trait(Traits.Feature, Traits.Features.Completion)]
+        [WorkItem(24490, "https://github.com/dotnet/roslyn/issues/24490")]
+        public async Task InParameter()
+        {
+            var markup = @"public abstract class Base
+{
+    public abstract void goo(int x, in string y);
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/OverrideCompletionProviderTests.cs]: ❗️ This needs to be in a different assembly to be a verification test for #24490.

Comment (jcouv) [reply src/EditorFeatures/CSharpTest/Completion/CompletionProviders/OverrideCompletionProviderTests.cs]: Ah, makes sense. Thanks

Comment (jcouv) [reply src/EditorFeatures/CSharpTest/Completion/CompletionProviders/OverrideCompletionProviderTests.cs]: Got a proper repro and updated PR

File: src/Workspaces/CSharp/Portable/CodeGeneration/ParameterGenerator.cs [hunk 2]
```diff
@@ -152,6 +152,7 @@ internal static ParameterSyntax GetParameter(IParameterSymbol p, CodeGenerationO
             {
                 return default;
             }
+            attributes = attributes.WhereAsArray(a => a.ToString() != "System.Runtime.CompilerServices.IsReadOnlyAttribute");
```
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/CodeGeneration/ParameterGenerator.cs]: 😨 A start would be `WellKnownTypes.GetMetadataName(WellKnownType.System_Runtime_CompilerServices_IsReadOnlyAttribute)`

❓ Why not move this to `AttributeGenerator.GenerateAttribute`? Returning null from that method suppresses the syntax.

💭 It makes me wonder if this should be handled more like `System.Runtime.CompilerServices.ExtensionAttribute`.

Comment (jcouv) [reply src/Workspaces/CSharp/Portable/CodeGeneration/ParameterGenerator.cs]: The reason I didn't move this filtering into `GenerateAttribute` is that it should only apply when the attribute is applied on a parameter. If it was applied somewhere else (for some reason), it should be printed, I think.

File: src/EditorFeatures/TestUtilities/Workspaces/TestWorkspaceFixture.cs [hunk 3]
```diff
@@ -19,6 +19,11 @@ public TestWorkspace GetWorkspace()
             return _workspace;
         }
 
+        public void SetWorkspace(TestWorkspace workspace)
```
Comment (jasonmalinowski) [inline src/EditorFeatures/TestUtilities/Workspaces/TestWorkspaceFixture.cs]: I think this would defeat the point of the fixture. This is to share a workspace between multiple tests. Mutating is uber-bad.

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/OverrideCompletionProviderTests.cs [hunk 4]
```diff
@@ -2693,5 +2694,60 @@ public class Class1
...
+{
+    public abstract void M(in int x);
+}";
+
+                var compilation = workspace.CurrentSolution.Projects.Single().GetCompilationAsync().GetAwaiter().GetResult();
+                var libComp = compilation.RemoveAllSyntaxTrees().AddSyntaxTrees(CSharpSyntaxTree.ParseText(markup, options: options));
+                var libRef = MetadataReference.CreateFromImage(Test.Utilities.CompilationExtensions.EmitToArray(libComp));
+
+                var project = workspace.CurrentSolution.Projects.Single();
+                var updatedProject = project.AddMetadataReference(libRef);
```
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/OverrideCompletionProviderTests.cs]: I think we have support for doing all this in the workspace already...? If not, might be better to add there.

File: src/EditorFeatures/TestUtilities/Workspaces/TestWorkspaceFixture.cs (hunk 3 above)
Comment (jcouv) [reply src/EditorFeatures/TestUtilities/Workspaces/TestWorkspaceFixture.cs]: I created an inner test class with a single test, so I think it's safe.
That said, I didn't like having to do that. The problem is that all the good test machinery relies on `WorkspaceFixture.GetWorkspace()` rather than that the `CreateWorkspace` in the test class :-(

Another approach I considered was creating a new derived class from `TestWorkspaceFixture`, but then I would have to re-implement a number of methods from `AbstractCSharpCompletionProviderTests` (since I couldn't derive from it anymore).

Not sure what a better design would be...

File: src/Workspaces/CSharp/Portable/CodeGeneration/ParameterGenerator.cs (hunk 2 above)
Comment (jcouv) [reply src/Workspaces/CSharp/Portable/CodeGeneration/ParameterGenerator.cs]: `WellKnownTypes` is inaccessible.

I looked around to see how `ExtensionAttribute` is handled, but couldn't find any relevant logic. Since extensions methods are static, they can't be overridden, so it's possible that the case is just not handled.


---
In reply to: [164312453](https://github.com/dotnet/roslyn/pull/24492#discussion_r164312453) [](ancestors = 164312453)

Comment (sharwell) [reply src/Workspaces/CSharp/Portable/CodeGeneration/ParameterGenerator.cs]: 💭 Since we already have `IParameterSymbol.RefKind`, it seems ideal to remove the attribute from the list of attributes provided by the API.

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for review. This PR now filters `IsReadOnly` from the parameter attributes. Thanks
The IDE change is just a test now.

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/OverrideCompletionProviderTests.cs (hunk 4 above)
Comment (jcouv) [reply src/EditorFeatures/CSharpTest/Completion/CompletionProviders/OverrideCompletionProviderTests.cs]: This is following the model of the previous test.

---
In reply to: [164599566](https://github.com/dotnet/roslyn/pull/24492#discussion_r164599566) [](ancestors = 164599566)

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs [hunk 5]
```diff
@@ -79,7 +79,7 @@ class Test
                 Assert.Equal(RefKind.RefReadOnly, method.RefKind);
                 Assert.True(method.ReturnsByRefReadonly);
 
-                AssertReferencedIsReadOnlyAttribute(Accessibility.Internal, method.GetReturnTypeAttributes(), module.ContainingAssembly.Name);
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs]: I've removed some validation of the attribute, which I intend to add back: verify it is emitted in metadata and that its accessibility is internal.

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.cs [hunk 6]
```diff
@@ -607,7 +607,9 @@ public override ImmutableArray<CSharpAttributeData> GetAttributes()
                     MightContainExtensionMethods ? AttributeDescription.CaseSensitiveExtensionAttribute : default,
                     out _,
                     // Filter out [Obsolete], unless it was user defined
-                    (IsByRefLikeType && ObsoleteAttributeData is null) ? AttributeDescription.ObsoleteAttribute : default);
+                    (IsByRefLikeType && ObsoleteAttributeData is null) ? AttributeDescription.ObsoleteAttribute : default,
+                    out _,
+                    default);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.cs]: >default [](start = 20, length = 7)

It feels like we should filter out IsReadonly attribute if ```IsReadOnly``` is true.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 5). Test changes are not looked at.

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_IsByRefLike.cs [hunk 7]
```diff
@@ -730,18 +753,10 @@ class Test
...
-                var assemblyName = module.ContainingAssembly.Name;
+                Assert.Equal(1, attributes.Count());
 
-                var attributeType = attributes[0].AttributeClass;
-                Assert.Equal("System.Runtime.CompilerServices.IsByRefLikeAttribute", attributeType.ToDisplayString());
-                Assert.Equal(assemblyName, attributeType.ContainingAssembly.Name);
-                Assert.Equal(accessibility, attributeType.DeclaredAccessibility);
-
-                var attribute = attributes[1];
+                var attribute = attributes.Single();
```
Comment (OmarTawfik) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_IsByRefLike.cs]: >attributes.Single(); [](start = 32, length = 20)

`Single()` throws if there isn't exactly one element. So I think the above assert is redundant.

File: src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs [hunk 8]
```diff
@@ -575,10 +575,10 @@ public void RefAssembly_ReferenceAssemblyAttributeAlsoInSource()
                 var reader = assembly.GetMetadataReader();
                 var attributes = reader.GetAssemblyDefinition().GetCustomAttributes();
                 AssertEx.Equal(new string[] {
-                        "MemberReference:Void System.Runtime.CompilerServices.CompilationRelaxationsAttribute.ctor(Int32)",
-                        "MemberReference:Void System.Runtime.CompilerServices.RuntimeCompatibilityAttribute.ctor()",
```
Comment (OmarTawfik) [inline src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs]: I don't see what source changes could have led to changing these strings.

Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs]: The extra dot comes from a fix in MetadataReaderUtils (line 322).

---
In reply to: [166739245](https://github.com/dotnet/roslyn/pull/24492#discussion_r166739245) [](ancestors = 166739245)

File: src/Test/Utilities/Portable/Metadata/MetadataReaderUtils.cs [hunk 9]
```diff
@@ -430,6 +431,34 @@ internal static void VerifyMethodBodies(ImmutableArray<byte> peImage, Action<byt
...
+            Assert.True(foundCount == 1, $"Got {foundCount}. Actual attributes were: {customAttributes.Select(a => Dump(reader, reader.GetCustomAttribute(a).Constructor)).Join(", ")}");
+
+            var isByRefLike = isByRefLikes.Single();
+            var attributeData = reader.GetCustomAttribute(isByRefLike);
+            var attributeCtor = reader.GetMethodDefinition((MethodDefinitionHandle)attributeData.Constructor);
+            var attribute = reader.GetTypeDefinition(attributeCtor.GetDeclaringType());
+            Assert.Equal(name, reader.GetString(attribute.Name));
+            Assert.Equal("System.Runtime.CompilerServices", reader.GetString(attribute.Namespace));
+
+            var accesibilityFlags = attribute.Attributes & TypeAttributes.VisibilityMask;
```
Comment (cston) [inline src/Test/Utilities/Portable/Metadata/MetadataReaderUtils.cs]: Typo: `accessibility`

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEModuleSymbol.cs [hunk 10]
```diff
@@ -290,34 +290,53 @@ internal void LoadCustomAttributes(EntityHandle token, ref ImmutableArray<CSharp
...
             out CustomAttributeHandle filteredOutAttribute2,
-            AttributeDescription filterOut2)
-        {
-            filteredOutAttribute1 = default(CustomAttributeHandle);
-            filteredOutAttribute2 = default(CustomAttributeHandle);
+            AttributeDescription filterOut2,
+            out CustomAttributeHandle filteredOutAttribute3,
+            AttributeDescription filterOut3,
+            out CustomAttributeHandle filteredOutAttribute4,
+            AttributeDescription filterOut4)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEModuleSymbol.cs]: The method comment looks stale.

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEModuleSymbol.cs [hunk 11]
```diff
@@ -290,34 +290,53 @@ internal void LoadCustomAttributes(EntityHandle token, ref ImmutableArray<CSharp
...
+                        filteredOutAttribute3 = customAttributeHandle;
+                        continue;
+                    }
+
+                    if (filterOut4.Signatures != null &&
+                        Module.GetTargetAttributeSignatureIndex(customAttributeHandle, filterOut4) != -1)
+                    {
+                        filteredOutAttribute4 = customAttributeHandle;
+                        continue;
+                    }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEModuleSymbol.cs]: Consider extracting a local helper method. Perhaps something like:
```
bool hasSignature(CustomAttributeHandle handle, AttributeDescription filter) =>
    filter.Signatures != null && Module.GetTargetAttributeSignatureIndex(handle, filter) != 1;
```

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs [hunk 12]
```diff
@@ -744,18 +744,25 @@ public override ImmutableArray<CSharpAttributeData> GetAttributes()
...
                     ImmutableArray<CSharpAttributeData> attributes =
                         containingPEModuleSymbol.GetCustomAttributesForToken(
                             _handle,
                             out paramArrayAttribute,
-                            filterOutParamArrayAttribute ? AttributeDescription.ParamArrayAttribute : default(AttributeDescription),
+                            filterOutParamArrayAttribute ? AttributeDescription.ParamArrayAttribute : default,
                             out constantAttribute,
-                            filterOutConstantAttributeDescription);
+                            filterOutConstantAttributeDescription,
+                            out  isReadOnlyAttribute,
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: Extra space.

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEPropertySymbol.cs [hunk 13]
```diff
@@ -532,7 +532,19 @@ public override ImmutableArray<CSharpAttributeData> GetAttributes()
...
+                ImmutableArray<CSharpAttributeData> attributes = containingPEModuleSymbol.GetCustomAttributesForToken(
+                      _handle,
+                      out _,
+                      this.RefKind == RefKind.RefReadOnly ? AttributeDescription.IsReadOnlyAttribute : default,
+                      out _,
+                      default,
+                      out _,
+                      default,
+                      out _,
+                      default);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEPropertySymbol.cs]: It looks like several uses of `GetCustomAttributesForToken()` take a single filter. Consider adding an overload for that case.

**Comment by jcouv** [comment]
@cston The last commit addresses your feedback. Thanks

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_ReadOnlyStruct.cs [hunk 14]
```diff
@@ -30,12 +30,19 @@ class Test
 }
 ";
 
-            CompileAndVerify(text, verify: Verification.Passes, symbolValidator: module =>
+            void validatePE(PEAssembly assembly)
+            {
+                var reader = assembly.GetMetadataReader();
+                var s1 = reader.TypeDefinitions.Select(tdh => reader.GetTypeDefinition(tdh)).Where(td => reader.GetString(td.Name) == "S1").Single();
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_ReadOnlyStruct.cs]: Perhaps add a `MetadataReaderUtils.GetTypeDefinitionByName()` helper method.

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs [hunk 15]
```diff
@@ -29,7 +32,19 @@ class Test
 }
 ";
 
-            CompileAndVerify(text, verify: Verification.Fails, symbolValidator: module =>
+            void validatePE(PEAssembly assembly)
+            {
+                var reader = assembly.GetMetadataReader();
+                var test = reader.TypeDefinitions.Select(tdh => reader.GetTypeDefinition(tdh)).Where(td => reader.GetString(td.Name) == "Test").Single();
+                var method = test.GetMethods().Select(m => reader.GetMethodDefinition(m)).Where(md => reader.GetString(md.Name) == "M").Single();
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs]: Add a `MetadataReaderUtils.GetMethodByName(MetadataReader, TypeDefinition, string)`?

File: src/Test/Utilities/Portable/Metadata/MetadataReaderUtils.cs [hunk 16]
```diff
@@ -430,6 +431,34 @@ internal static void VerifyMethodBodies(ImmutableArray<byte> peImage, Action<byt
 
         static readonly byte[] ThrowNull = new[] { (byte)ILOpCode.Ldnull, (byte)ILOpCode.Throw };
 
+        internal static void AssertReferencedIsReadOnlyAttributes(MetadataReader reader, CustomAttributeHandleCollection customAttributes, Accessibility accessibility)
```
Comment (cston) [inline src/Test/Utilities/Portable/Metadata/MetadataReaderUtils.cs]: Consider making these extension methods.

File: src/Test/Utilities/Portable/Metadata/MetadataReaderUtils.cs [hunk 17]
```diff
@@ -430,6 +431,34 @@ internal static void VerifyMethodBodies(ImmutableArray<byte> peImage, Action<byt
...
+        }
+
+        internal static void AssertReferencedIsByRefLikeAttributes(MetadataReader reader, CustomAttributeHandleCollection customAttributes, Accessibility accessibility)
+        {
+            AssertAttributePresent(reader, customAttributes, accessibility, "IsByRefLikeAttribute");
+        }
+
+        private static void AssertAttributePresent(MetadataReader reader, CustomAttributeHandleCollection customAttributes, Accessibility accessibility, string name)
+        {
+            var isByRefLikes = customAttributes.Where(a => Dump(reader, reader.GetCustomAttribute(a).Constructor) == $"MethodDefinition:Void {name}..ctor()");
```
Comment (cston) [inline src/Test/Utilities/Portable/Metadata/MetadataReaderUtils.cs]: Not necessarily `IsByRefLike`.

File: src/Test/Utilities/Portable/Metadata/MetadataReaderUtils.cs [hunk 18]
```diff
@@ -15,6 +15,7 @@
 using Microsoft.CodeAnalysis.Text;
 using System.IO;
 using System.IO.Compression;
+using Xunit;
```
Comment (cston) [inline src/Test/Utilities/Portable/Metadata/MetadataReaderUtils.cs]: Should not be needed in `MetadataReaderUtils`.

File: src/Test/Utilities/Portable/Metadata/MetadataReaderUtils.cs (hunk 16 above)
Comment (jcouv) [reply src/Test/Utilities/Portable/Metadata/MetadataReaderUtils.cs]: The advantage of asserting in those methods is that the test failure will give more specific information about which part of the validation failed, and while debugging it will stop at the right place too.

---
In reply to: [167377740](https://github.com/dotnet/roslyn/pull/24492#discussion_r167377740) [](ancestors = 167377740)

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_IsByRefLike.cs [hunk 19]
```diff
@@ -29,11 +31,22 @@ class Test
 }
 ";
 
-            CompileAndVerify(text, verify: Verification.Passes, symbolValidator: module =>
+
+            void validatePE(PEAssembly assembly)
+            {
+                MetadataReader reader = assembly.GetMetadataReader();
+                var s1 = reader.TypeDefinitions.Select(tdh => reader.GetTypeDefinition(tdh)).Where(td => reader.GetString(td.Name) == "S1").Single();
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_IsByRefLike.cs]: `GetTypeDefinitionByName` here and in several other places.

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs [hunk 20]
```diff
@@ -29,7 +32,19 @@ class Test
 }
 ";
 
-            CompileAndVerify(text, verify: Verification.Fails, symbolValidator: module =>
+            void validatePE(PEAssembly assembly)
+            {
+                var reader = assembly.GetMetadataReader();
+                var test = reader.GetTypeDefinitionByName("Test");
+                var method = reader.GetMethodDefinitionByName(test, "M");
+                var returnType = reader.GetParameter(method.GetParameters().ElementAt(1));
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs]: `var returnType = ...ElementAt(0);`?

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadonlyStructTests.cs [hunk 21]
```diff
@@ -817,6 +817,14 @@ public override string ToString()
             Assert.Equal(RefKind.In, namedType.GetMethod("M1").ThisParameter.RefKind);
             Assert.Equal(RefKind.In, namedType.GetMethod("ToString").ThisParameter.RefKind);
 
+            void validatePE(PEAssembly assembly)
+            {
+                var reader = assembly.GetMetadataReader();
+                var test = reader.TypeDefinitions.Select(tdh => reader.GetTypeDefinition(tdh)).Where(td => reader.GetString(td.Name) == "S1").Single();
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadonlyStructTests.cs]: `GetTypeDefinitionByName`

File: src/Test/Utilities/Portable/Metadata/MetadataReaderUtils.cs [hunk 22]
```diff
@@ -428,6 +428,45 @@ internal static void VerifyMethodBodies(ImmutableArray<byte> peImage, Action<byt
...
+        {
+            return reader.ContainsAttribute(customAttributes, accessibility, "IsReadOnlyAttribute");
+        }
+
+        internal static bool ContainsIsByRefLikeAttributes(this MetadataReader reader, CustomAttributeHandleCollection customAttributes, Accessibility accessibility)
+        {
+            return reader.ContainsAttribute(customAttributes, accessibility, "IsByRefLikeAttribute");
+        }
+
+        private static bool ContainsAttribute(this MetadataReader reader, CustomAttributeHandleCollection customAttributes, Accessibility accessibility, string name)
```
Comment (AlekseyTs) [inline src/Test/Utilities/Portable/Metadata/MetadataReaderUtils.cs]: >MetadataReader [](start = 51, length = 14)

I think it would be better to use PEModule instance and rely on helpers within this type to find attributes based on specific AttributeDescription

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs (hunk 20 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs]: Good catch. Thanks

---
In reply to: [168014179](https://github.com/dotnet/roslyn/pull/24492#discussion_r168014179) [](ancestors = 168014179)

**Comment by jcouv** [comment]
@jasonmalinowski Ok with the test change?
Also, how do I unblock the WIP bot? I've already removed WIP from the title...

**Comment by sharwell** [comment]
@jcouv It's because you have a commit (58d29bd63e9e8b191b8f9608a12d8d6ca21a4ffe) with WIP in the commit message.

File: src/Test/Utilities/Portable/Metadata/MetadataReaderUtils.cs (hunk 22 above)
Comment (jcouv) [reply src/Test/Utilities/Portable/Metadata/MetadataReaderUtils.cs]: @AlekseyTs One downside of the helpers (`HasIsReadOnlyAttribute`, and the IsByRef one) is that they don't help check the accessibility. Are you ok with relaxing our tests that way?

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEModuleSymbol.cs [hunk 23]
```diff
@@ -535,7 +536,8 @@ internal NamedTypeSymbol SystemTypeSymbol
             }
         }
 
-        private NamedTypeSymbol GetTypeSymbolForWellKnownType(WellKnownType type)
+        // internal for testing purpose
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEModuleSymbol.cs]: >// [](start = 8, length = 2)

Please use standard doc comment, so that intellisense would show that

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_IsByRefLike.cs [hunk 24]
```diff
@@ -29,11 +31,26 @@ class Test
...
+            }
+
+            void validate(ModuleSymbol module)
             {
                 var type = module.ContainingAssembly.GetTypeByMetadataName("Test").GetTypeMember("S1");
-                AssertReferencedIsByRefLikeAttributes(Accessibility.Public, type, module.ContainingAssembly.Name);
-            });
+                AssertReferencedIsByRefLike(type);
+
+                if (module is PEModuleSymbol peModule)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_IsByRefLike.cs]: >module is PEModuleSymbol peModule [](start = 20, length = 33)

Is this ever false?

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_IsByRefLike.cs [hunk 25]
```diff
@@ -29,11 +31,26 @@ class Test
 }
 ";
 
-            CompileAndVerify(text, verify: Verification.Passes, symbolValidator: module =>
+
+            void validatePE(PEAssembly assembly)
+            {
+                MetadataReader reader = assembly.GetMetadataReader();
+                var s1 = reader.GetTypeDefinitionHandleByName("S1");
+                Assert.True(assembly.Modules.Single().HasIsByRefLikeAttribute(s1));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_IsByRefLike.cs]: It is up to you if you want to simplify this, but it all can be done much simpler in the ```validate``` below:
```
Assert.True(((PEModuleSymbol)module).Module.HasIsByRefLikeAttribute(((PENamedTypeSymbol)type).Handle));
```

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_IsByRefLike.cs [hunk 26]
```diff
@@ -935,7 +951,7 @@ void Method()
         public void SpecialTypes_CorLib()
         {
             var source1 =
-@"
+    @"
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_IsByRefLike.cs]: Is this change intentional?

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs [hunk 27]
```diff
@@ -29,7 +32,19 @@ class Test
 }
 ";
 
-            CompileAndVerify(text, verify: Verification.Fails, symbolValidator: module =>
+            void validatePE(PEAssembly assembly)
+            {
+                var reader = assembly.GetMetadataReader();
+                var test = reader.GetTypeDefinitionHandleByName("Test");
+                var method = reader.GetMethodDefinitionByName(reader.GetTypeDefinition(test), "M");
+                var returnTypeHandle = method.GetParameters().ElementAt(0);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs]: >method.GetParameters().ElementAt(0) [](start = 39, length = 35)

Symbol for return parameter is available through PEMethodSymbol.Signature.ReturnParam

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs [hunk 28]
```diff
@@ -29,7 +32,19 @@ class Test
...
+                var parameterHandle = method.GetParameters().ElementAt(1);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs]: >ar parameterHandle = method.GetParameters().ElementAt(1); [](start = 17, length = 57)

Similarly, parameter handles are available from PEParameterSymbol

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs (hunk 27 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs]: Thanks. That is indeed much nicer. I assume it's ok for me to make `Signature` and its return type `SignatureData` internal for testing purpose.

---
In reply to: [168326601](https://github.com/dotnet/roslyn/pull/24492#discussion_r168326601) [](ancestors = 168326601)

**Comment by jcouv** [comment]
Squashed the commits to solve issue with WIP bot.
@jaredpar for ask-mode approval for 15.7. Thanks

**Comment by jaredpar** [comment]
approved

**Comment by jcouv** [comment]
ubuntu_16_mono_debug_prtest failed with nuget error ([details](https://ci.dot.net/job/dotnet_roslyn/job/dev15.7.x/job/ubuntu_16_mono_debug_prtest/293/))

@dotnet/roslyn-infrastructure FYI I'll re-run.

```39;49m[31m/opt/code/Binaries/Tools/dotnet/sdk/2.2.0-preview1-007622/NuGet.targets(103,5): error : Failed to retrieve information about 'Microsoft.VisualBasic' from remote source 'https://api.nuget.org/v3-flatcontainer/microsoft.visualbasic/index.json'. [/opt/code/Compilers.sln]```

**Comment by jcouv** [comment]
test ubuntu_16_mono_debug_prtest  please

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_IsByRefLike.cs [hunk 29]
```diff
@@ -29,11 +31,17 @@ class Test
...
+            void validate(ModuleSymbol module)
             {
                 var type = module.ContainingAssembly.GetTypeByMetadataName("Test").GetTypeMember("S1");
-                AssertReferencedIsByRefLikeAttributes(Accessibility.Public, type, module.ContainingAssembly.Name);
-            });
+                AssertReferencedIsByRefLike(type);
+
+                var peModule = (PEModuleSymbol)module;
+                Assert.True(peModule.Module.HasIsByRefLikeAttribute(((PENamedTypeSymbol)type).Handle));
+                Assert.Equal(Accessibility.Public, peModule.GetTypeSymbolForWellKnownType(WellKnownType.System_Runtime_CompilerServices_IsByRefLikeAttribute).DeclaredAccessibility);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_IsByRefLike.cs]: >GetTypeSymbolForWellKnownType [](start = 60, length = 29)

I just realized that this helper is probably not quite what we should use here. In particular it does not guarantee that the type it finds is defined in this module. It feels like an existing test helper that looks for a type in the given container (module's global namespace in our case) is a better fit. I think there is one that takes a fully qualified name.

**Comment by jcouv** [comment]
windows_debug_unit64_prtest  failed with test timeout ([details](https://ci.dot.net/job/dotnet_roslyn/job/dev15.7.x/job/windows_debug_unit64_prtest/291/))

I'll re-run.
`Roslyn Error: test timeout exceeded, dumping remaining processes`

**Comment by jcouv** [comment]
test windows_debug_unit64_prtest please

**Comment by jcouv** [comment]
test windows_release_unit64_prtest  please

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_IsByRefLike.cs (hunk 29 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_IsByRefLike.cs]: Thanks
Narrowed the check. I couldn't find an existing helper, so followed the pattern used a bunch of NoPia tests (GetMetadataName and use it in Lookup method).

---
In reply to: [168557695](https://github.com/dotnet/roslyn/pull/24492#discussion_r168557695) [](ancestors = 168557695)

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_IsByRefLike.cs]: >  I couldn't find an existing helper

src\Compilers\Test\Utilities\CSharp\Extensions.cs
```
public static T GetMember<T>(this NamespaceOrTypeSymbol symbol, string qualifiedName) where T : Symbol
        {
            return (T)symbol.GetMember(qualifiedName);
        }
```

---
In reply to: [168574351](https://github.com/dotnet/roslyn/pull/24492#discussion_r168574351) [](ancestors = 168574351,168557695)

File: src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs [hunk 30]
```diff
@@ -871,6 +871,11 @@ internal IEnumerable<string> GetAttributeNames(ImmutableArray<CSharpAttributeDat
             return attributes.Select(a => a.AttributeClass.Name);
         }
 
+        internal void AssertHasAttribute(PEModuleSymbol peModule, WellKnownType type, Accessibility expectedAccessibility)
```
Comment (AlekseyTs) [inline src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs]: >AssertHasAttribute [](start = 22, length = 18)

AssertDeclaresType? Is there anything specific for attributes here?

**Comment by jcouv** [comment]
@AlekseyTs I renamed the test helper. Good to go?
